### ISSUE TITLE
macOS: Remember last focused window location for next startup

### ISSUE THREAD

**Issue by mitchellh** [ghostty-org/ghostty#4233]
### Discussed in https://github.com/ghostty-org/ghostty/discussions/4204

<div type='discussions-op-text'>

<sup>Originally posted by **evertonstz** December 31, 2024</sup>
I'm one of those guys that keeps closing and reopening their terminal during the day and that's one of the first things I noticed when using Ghostty: it doesn't remembers the size and position of the last closed window like most other MacOS apps.
It's been a long time since I used Linux with modern gnome, but I think that's also the expected behavior for apps there, right?</div>

## Research on Other Behavior

`Terminal.app` and `iTerm` both remember the position of the last focused window (not last created window) and restore it when restarting the app.  Only the position is remembered, not the size.

**Note this is different from `window-save-state`.** This isn't _restoring_ any window. This is just remembering the _position_.

This behavior isn't typical with Gnome applications, so this is tagged macOS only.

## Implementation

Given this is a macOS only behavior, I'm inclined to leverage macOS APIs for this and put this information into `defaults`. I suspect that's where these programs are also doing it but verification required.

I think we should solve this completely in the Swift code.

There are no configs currently to specify an initial window position, but in the future if/when we have those, those should always override this.

**Comment by huhu415** [comment]
My config file: `window-save-state = always` and `quit-after-last-window-closed=true`
I also found that if you use `comand + q` to exit ghostty, the next time you start it, the position and size will be remembered. 

But if I use `comand + w` to exit ghostty, even if I set `quit-after-last-window-closed=true`, even though it has already exited, the next time I open it, I still won't remember the setting and size

**Comment by jsumners** [comment]
I look forward to this being implemented. I tend to keep my terminal in the lower left corner of my workspace. I also utilize multiple "Spaces," usually one per project. So I end up creating multiple terminal windows, at least one per Space, and, while not a deal breaker, it gets tiresome having to manually move the new window to the location I typically keep it.

**Comment by karlhorky** [comment]
@jsumners does the `window-save-state` config setting work for you?

```ini
window-save-state = always
```

> `window-save-state`
>
> Whether to enable saving and restoring window state. Window state includes their position, size, tabs, splits, etc.

> - `always` will always save window state whenever Ghostty is exited.

This works for me so far.

**Comment by mitchellh** [comment]
Note this issue is different from `window-save-state`. That remembers windows that were _open_. The behavior being described in this issue is that macOS apps appear to remember the last location (not size) of windows that were focused or closed in order to open the _first_ window on the next app startup.

**Comment by jsumners** [comment]
> Note this issue is different from `window-save-state`. That remembers windows that were _open_. The behavior being described in this issue is that macOS apps appear to remember the last location (not size) of windows that were focused or closed in order to open the _first_ window on the next app startup. 

Right. And I'm adding more information that the last focused window should _also_ inform the position of the next new window created from the Dock icon's context menu.

**Comment by pluiedev** [comment]
Note that on a technically level we *cannot* make this work on Linux, at least not just for Ghostty. GTK and Wayland explicitly do not allow windows to move themselves, and the placement of each window is solely up to the discretion of the compositor/window manager, and this functionality can only be implemented there.

**Comment by rizalstrato** [comment]
shell-integration = zsh
window-save-state = always
quit-after-last-window-closed=true

i used these three on config file. (i used zsh)
Worked everytime

**Comment by markwyner** [comment]
Thanks for this information, @mitchellh. Do you know where this fits for release on your roadmap? It's one of only two things about Ghostty that are really bothering me. It's an otherwise amazing terminal. Excellent work overall.

**Comment by jsumners** [comment]
Nothing can be released if no one PRs a fix. I do not see any PRs associated with this issue. I'm sure they'd welcome your contribution.

**Comment by atticusmatticus** [comment]
> My config file: `window-save-state = always` and `quit-after-last-window-closed=true` I also found that if you use `comand + q` to exit ghostty, the next time you start it, the position and size will be remembered.
> 
> But if I use `comand + w` to exit ghostty, even if I set `quit-after-last-window-closed=true`, even though it has already exited, the next time I open it, I still won't remember the setting and size

Thank you for this suggestion. Similarly, if exiting the terminal with `control + d` the window position is not remembered either.

**Comment by quanru** [comment]
> > My config file: `window-save-state = always` and `quit-after-last-window-closed=true` I also found that if you use `comand + q` to exit ghostty, the next time you start it, the position and size will be remembered.
> > But if I use `comand + w` to exit ghostty, even if I set `quit-after-last-window-closed=true`, even though it has already exited, the next time I open it, I still won't remember the setting and size
> 
> Thank you for this suggestion. Similarly, if exiting the terminal with `control + d` the window position is not remembered either.

What method can make cmd + w remember the window state after closing?

**Comment by abdurrahmanski** [comment]
Has anyone found a solution for making Command+W remember the window position and size like Command+Q does? (Commenting to see if this is still being tracked since the issue was closed)
